modules/contracts/src.ts/services/ethReader.ts

```diff
@@ -574,40 +558,61 @@ export class EthereumChainReader implements IVectorChainReader {
     chainId: number,
     bytecode?: string,
   ): Promise<Result<RegisteredTransfer[], ChainError>> {
+    return await this.retryWrapper<RegisteredTransfer[]>(chainId,
+      async (provider: JsonRpcProvider) => {
+        // Registry for chain not loaded, load into memory
+        const registry = new Contract(transferRegistry, TransferRegistry.abi, provider);
+        let registered;
+        if (bytecode) {
+          // Try with evm first
+          const evm = this.tryEvm(registry.interface.encodeFunctionData("getTransferDefinitions"), bytecode);
+
+          if (!evm.isError) {
+            try {
+              registered = registry.interface.decodeFunctionResult("getTransferDefinitions", evm.getValue()!)[0];
+            } catch (e) {}
+          }
+        }
+        if (!registered) {
+          try {
+            registered = await registry.getTransferDefinitions();
+          } catch (e) {
+            return Result.fail(new ChainError(e.message, { chainId, transferRegistry }));
+          }
+        }
+        const cleaned = registered.map((r: RegisteredTransfer) => {
+          return {
+            name: r.name,
+            definition: r.definition,
+            stateEncoding: tidy(r.stateEncoding),
+            resolverEncoding: tidy(r.resolverEncoding),
+            encodedCancel: r.encodedCancel,
+          };
+        });
+        this.transferRegistries.set(chainId.toString(), cleaned);
+        return Result.ok(cleaned);
+      }
+    );
+  }
+
+  private async retryWrapper<T>(
+    chainId: number,
+    targetMethod: (provider: JsonRpcProvider) => Promise<Result<T, ChainError>>
+  ): Promise<Result<T, ChainError>> {
+    // TODO : Should we retry this attempt to get provider as well?
```

Comment:
No we can leave this outside of the retries

Reply:
Copy